File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第2章回形取数.py
#回形取数    lanqiaoOJ 1517

dir = [(1, 0), (0, 1), (-1, 0), (0, -1)]  #4个方向
m, n = map(int, input().split())
a = []
for i in range(m):   a.append(input().split())
x, y = -1, 0
d = 0
sum = 0
while sum < m*n:
    sum = sum + 1
    nx, ny = x + dir[d][0], y + dir[d][1]
    if nx < 0 or nx >= m or ny < 0 or ny >= n or a[nx][ny]==-1:
        d = (d + 1) % 4
        x, y = x + dir[d][0], y + dir[d][1]
    else:
        x, y = nx, ny
    print(a[x][y], end=' ')
    a[x][y] = -1             #标记这个坐标点已经取过

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第2章回形取数.java
//回形取数    lanqiaoOJ 1517

import java.util.Scanner;
public class Main {
static int[][] dir = new int[][] {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int m = sc.nextInt(), n = sc.nextInt();
        int[][] a = new int[m][n];
        for(int i=0; i<m; ++i)
            for(int j=0; j<n; ++j)
                a[i][j] = sc.nextInt();

        int x = -1, y = 0, d = 0, sum = 0;
        while(sum < m*n) {
            sum ++;
            int nx = x+dir[d][0], ny = y+dir[d][1];
            if(nx<0 || ny<0 || nx>=m || ny>=n || a[nx][ny]==-1) {
                d = (d+1)%4;
                x = x+dir[d][0];
                y = y+dir[d][1];
            }else {
                x = nx;
                y = ny;
            }
            System.out.print(a[x][y]+" ");
            a[x][y] = -1;    //标记这个坐标点已经取过
        }
    }
}
